sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldHaveCorrectSignature.cs

```diff
@@ -0,0 +1,111 @@
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class TestMethodShouldHaveCorrectSignature : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3433";
+        private const string MessageFormat = "Make this test method {0}.";
+        private const string MakePublicMessage = "'public'";
+        private const string MakeNonAsyncOrTaskMessage = "non-'async' or return 'Task'";
+        private const string MakeNotGenericMessage = "non-generic";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<KnownType> TrackedTestAttributes = ImmutableHashSet.Create(
+            KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_TestMethodAttribute,
+            KnownType.NUnit_Framework_TestAttribute,
+            KnownType.Xunit_FactAttribute);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    if (!c.IsTest())
+                    {
+                        return;
+                    }
+
+                    var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                    var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+                    if (classSymbol == null)
+                    {
+                        return;
+                    }
+
+                    var allFaultyMethods = classSymbol.GetMembers()
+                        .OfType<IMethodSymbol>()
+                        .Where(IsTestMethod)
+                        .Select(method => new
+                        {
+                            Location = method.Locations.First(),
+                            Message = GetFaults(method).ToSentence()
+                        })
+                        .Where(tuple => tuple.Message != null);
+
+                    foreach (var faultyMethod in allFaultyMethods)
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, faultyMethod.Location,
+                            faultyMethod.Message));
+                    }
+                },
+                SyntaxKind.ClassDeclaration);
+        }
+
+        private static bool IsTestMethod(IMethodSymbol method)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Consider using more general type 'Microsoft.CodeAnalysis.ISymbol' instead of 'Microsoft.CodeAnalysis.IMethodSymbol'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3242)